Tangent Interface Docs

willtebbutt opened the pull request:
<!--
    Thank you for opening a pull request to Mooncake.jl!
    Please note that this project operates the following policy: any time a PR is merged
    which modifies code in the src directory, a release must be made.

    Consequently, if your PR modifies something in src, please follow semver to figure out
    how you should modify the version number in Project.toml. If it's unclear to you how the
    version number should be modified, please open your PR without modifying it, and ask for
    assistance -- one of the maintainers will be happy to help figure out what the new
    version number ought to be.
-->
@jollywatt per our discussion elsewhere, here is a stab at aggregating all of the information that I have about the tangent interface into a single location. When you get a minute, could you let me know if this is the kind of thing that you need to do what you wanted to do, or if there's additional info that you'll need?

sunxd3 commented on docs/src/developer_documentation/tangents.md:
```diff
@@ -0,0 +1,47 @@
+# Tangents
+
+As discussed in [Representing Gradients](@ref), Mooncake requires that each "primal" type be associated to a unique "tangent" type, given by the function [tangent_type](@ref).
+Moreover, we must be able to "split" a given tangent into its _fdata_ ("forwards-data") and _rdata_ ("reverse-data"), whose types are given by [fdata_type](@ref) and `rdata_type` respectively.
```
maybe miss `ref` for `rdata_type`?

willtebbutt replied:
Ah, yes. For some reason, it didn't like me doing that. I'm not 100% sure why.

sunxd3 commented on src/test_utils.jl:
```diff
@@ -765,15 +765,22 @@ function functions_for_mutable_structs()
 
 """
-    test_rule_and_type_interactions(rng::AbstractRNG, x)
+    test_rule_and_type_interactions(rng::AbstractRNG, p)
```
(not for this line) in the first sentence of the following doc, there is an `x` possibly should be `p` (also the first sentence of the second paragraph)?

willtebbutt replied:
Good catch, thanks!

sunxd3 reviewed (commented):
I haven't tried to implement a tangent type and use these functions, so couldn't verify claims like `If all the tests in these functions pass, then you have satisfied the interface.` All the comments are only cosmetic. 

But this is clearly a really good clarification!

willtebbutt commented on src/test_utils.jl:
```diff
@@ -765,15 +765,22 @@ function functions_for_mutable_structs()
+    test_rule_and_type_interactions(rng::AbstractRNG, p)
 
 Check that a collection of standard functions for which we _ought_ to have a working rrule
 for `x` work, and produce the correct answer. For example, the `rrule!!` for `typeof` should
```
```suggestion
for `p` work, and produce the correct answer. For example, the `rrule!!` for `typeof` should
```

willtebbutt commented on src/test_utils.jl:
```diff
@@ -765,15 +765,22 @@ function functions_for_mutable_structs()
 struct-type, and we should have a rule for `setfield!` for any mutable struct type.
+See implementation for details.
 
 The purpose of this test is to ensure that, for any given `x`, the full range of primitive
```
```suggestion
The purpose of this test is to ensure that, for any given `p`, the full range of primitive
```

Jollywatt commented:
@willtebbutt I've had a look at these docs and tried to implement the interface for a custom type. I was almost able to, but in particular I had to add methods for
- `_get_fdata_field`
- `tangent` (binary method)

in addition to those in the interface. I also ran into problems with a generated function, which I didn't find a way around.

I've put a MWE in this notebook: https://jollywatt.github.io/notes/mooncake-self-tangents

willtebbutt commented:
Thanks -- I'll take a look at your notebook!

willtebbutt commented:
@Jollywatt sorry for taking so long to get this fixed. I believe I've addressed the issues that you mentioned above.

In particular, I've added the binary version of `tangent` to the docstring for the tangent splitting tests, and stated in the docstring from the basic rules tests that you need to support `_new_` / `getfield` / `setfield!` etc (which is where the `_get_fdata_field` thing you mentioned is coming from).

Could you let me know if you think this is sufficient? If so, I'll merge this.

willtebbutt commented:
Ah, yes, you're definitely meant to test on the `<function>_internal` functions, but I've not updated this in the various docstrings 😬 . I forgot that I would need to reflect this change in them. I'll update them and let you know once that's done.

willtebbutt commented:
@Jollywatt I think that the docstring should now be accurate.

willtebbutt commented:
Thanks for the approval @yebai -- I'm going to wait for @Jollywatt to have another go at this before merging, as doing so seems to have caught issues quite reliably!

yebai commented:
I am happy to merge this PR and fix further issues (if any) in future PRs.

willtebbutt commented:
I'm still discussing this with @Jollywatt via email, so would prefer to keep it un-merged for now. We're nearly there, just ironing out the last couple of issues.

yebai commented:
As mentioned above, having these docs is helpful. Inaccuracies (if any) can be fixed via future PRs.
